### ISSUE TITLE
Desktop notifications for user-defined criteria

### ISSUE THREAD

**Issue by Chris-Bitler** [danreeves/dt-exchange#19]
Users who leave the page open should be able to, if they give the extension permission, get desktop notifications if an item meeting criteria they defined has appeared in the shop inventory(s).

Example filter criteria:
- Level
- Base Level
- Blessing
- Perks
- Modifier values?

Base level, blessings, and perks would be a good start atleast.

**Comment by errnoh** [comment]
👍 will likely be next on the line for me after the shop filter branch gets merged

**Comment by Chris-Bitler** [comment]
> 👍 will likely be next on the line for me after the shop filter branch gets merged

I've actually already started work on this, if that is alright?

@danreeves Please assign this issue to me to prevent any confusion

**Comment by errnoh** [comment]
@Chris-Bitler oh, well. I don't mind not having to work on the notification, but the trigger logic will likely have quite large overlap with #2, which was the main reason I was planning on working on that.

**Comment by danreeves** [comment]
you're right, i don't know if this is ready to be worked on beyond "the ability to create notifications, generically"

**Comment by Chris-Bitler** [comment]
@danreeves @errnoh I can base this issue's work on #2 if that is the case - I think it should be fine to start working on it, and I can update it with changes based on what happens with @errnoh's PR. Is that OK?

**Comment by tenstan** [comment]
Seems like this feature request has stranded for quite a while now, so I'm guessing that nobody is looking at this anymore. @danreeves I'm willing to take a crack at this. Do you still accept new features? Just wanna verify before I spend time on this, since it's been a while since a new feature has been added.

I'm curious about how we wanna implement this. Now that filters are implemented, I'd say a simple bell-like button would suffice to enable desktop notifications. You can press this button when `Enable rule based filtering` is toggled on. Whenever the shop refreshes, the items in the shop will be checked based on the configured filters, and if there is at least one match, then a desktop notification can be sent.

One potential issue I see is how we are going to make the browser do a periodic check. Simply requiring the page to be left open, might cause some issues where the browser suspends the page if it is inactive for long enough. I'm not sure yet if a browser extension can run background tasks or something alike, otherwise that could be a solution. We might also have to deal with token expiration to access the shop API in some way if a background task would be a solution.

**Comment by danreeves** [comment]
@tenstan Yeah, I'm open to PRs. I've struggled to find time for this. Background checks are certainly possible but my knowledge is more around v2 extensions, but chrome now forces v3 and I'm not too familiar with the APIs available yet

**Comment by tenstan** [comment]
Been looking at this for the past few days, but there's a pretty big obstacle in my eyes.

When authenticating, I get the following access token and refresh token from Fatshark:

|                     | Issued At               | Expires At              | Lifetime |
| ------------- | ----------------------- | ----------------------- | -------- |
| Access Token  | July 15th 2025 12:16:16 | July 15th 2025 12:57:16 | 00:41:00 |
| Refresh Token | July 15th 2025 12:16:16 | July 15th 2025 12:58:16 | 00:42:00 |

Now that's a VERY short lifetime for a refresh token, even for one that is stored locally. That explains why I had to re-authenticate so often while testing the extension at least.

I decided to check again, this time with different results:

|               | Issued At               | Expires At              | Lifetime |
| ------------- | ----------------------- | ----------------------- | -------- |
| Access Token  | July 15th 2025 12:27:13 | July 15th 2025 13:04:13 | 00:37:00 |
| Refresh Token | July 15th 2025 12:27:13 | July 15th 2025 13:05:13 | 00:38:00 |

So not only is the refresh token very short, both tokens also vary in lifetimes between renewals.

This makes it pretty tough to implement desktop notifications in a proper way. A background script that does a periodic check at the shop would make a lot of sense here, but it'll be very annoying to use if you are forced to re-authenticate every time you choose to leave your PC even for only less than an hour.

**Comment by tenstan** [comment]
The only way I see getting around it is by instructing the user to leave the page open if they want to continue to receiving notifications. At that point you may as well throw a background script out the window.

Though the user will have to ensure that the page doesn't get unloaded by the browser, otherwise the periodic check won't work. And if you've reached that point... Then you could just show a simple banner on the page with the current status instead of implementing desktop notifications.

The alternative would be reverse engineering how Fatshark performs auth on the front-end. We can add a button to the extension inside the browser toolbar, then make the user re-authenticate through there and pass the tokens to the background script, which... may be feasible
